aptos-move/move-examples/fungible_asset/stablecoin/sources/usdk.move

```diff
@@ -0,0 +1,325 @@
+            usdk_signer,
+            string::utf8(b"usdk"),
+            string::utf8(b"withdraw"),
+        );
+        dispatchable_fungible_asset::register_dispatch_functions(
+            constructor_ref,
+            option::some(withdraw),
+            option::some(deposit),
+            option::none(),
+        );
+    }
+
+    /// Allow a spender to transfer tokens from the owner's account given their signed approval.
+    /// Caller needs to provide the from account's scheme and public key which can be gotten via the Aptos SDK.
+    public fun transfer_from(
+        spender: &signer,
+        proof: vector<u8>,
+        from: address,
+        from_account_scheme: u8,
+        from_public_key: vector<u8>,
+        to: address,
+        amount: u64,
+    ) acquires Management, State {
+        assert_not_paused();
+        assert_not_denylisted(from);
+        assert_not_denylisted(to);
+
+        let expected_message = Approval {
+            owner: from,
+            nonce: account::get_sequence_number(from),
+            chain_id: chain_id::get(),
+            spender: signer::address_of(spender),
+            amount,
+        };
+        account::verify_signed_message(from, from_account_scheme, from_public_key, proof, expected_message);
+
+        let transfer_ref = &borrow_global<Management>(usdk_address()).transfer_ref;
+        primary_fungible_store::transfer_with_ref(transfer_ref, from, to, amount);
+    }
+
+    /// Deposit function override to ensure that the account is not denylisted and the stablecoin is not paused.
+    public fun deposit<T: key>(
+        store: Object<T>,
+        fa: FungibleAsset,
+        transfer_ref: &TransferRef,
+    ) acquires State {
+        assert_not_paused();
+        assert_not_denylisted(object::owner(store));
+        fungible_asset::deposit_with_ref(transfer_ref, store, fa);
+    }
+
+    /// Withdraw function override to ensure that the account is not denylisted and the stablecoin is not paused.
+    public fun withdraw<T: key>(
+        store: Object<T>,
+        amount: u64,
+        transfer_ref: &TransferRef,
+    ): FungibleAsset acquires State {
+        assert_not_paused();
+        assert_not_denylisted(object::owner(store));
+        fungible_asset::withdraw_with_ref(transfer_ref, store, amount)
+    }
+
+    /// Mint new tokens to the specified account. This checks that the caller is a minter, the stablecoin is not paused,
+    /// and the account is not denylisted.
+    public entry fun mint(minter: &signer, to: address, amount: u64) acquires Management, Roles, State {
+        assert_not_paused();
+        assert_is_minter(minter);
+        assert_not_denylisted(to);
+
+        let management = borrow_global<Management>(usdk_address());
+        let tokens = fungible_asset::mint(&management.mint_ref, amount);
+        deposit(primary_fungible_store::ensure_primary_store_exists(to, metadata()), tokens, &management.transfer_ref);
```

Comment:
To generalize this contract, later, might be worth passing in the address of the metadata in the function, i.e. passing in `Object<Metadata>` and having relevant admin checks in the contract.